Source/Core/DolphinWX/GameListCtrl.cpp

```diff
@@ -631,96 +668,148 @@ void GameListCtrl::SetColors()
+  }
+  else
+  {
+    std::vector<u8> buffer(f.GetSize());
+    if (buffer.size() && f.ReadBytes(buffer.data(), buffer.size()))
     {
-      auto gli = std::make_unique<GameListItem>(drive, title_database);
-
-      if (gli->IsValid())
-        m_ISOFiles.push_back(std::move(gli));
+      u8* ptr = buffer.data();
+      PointerWrap p(&ptr, PointerWrap::MODE_READ);
+      DoState(&p, buffer.size());
+      if (p.GetMode() == PointerWrap::MODE_READ)
+        success = true;
     }
   }
+  if (!success)
+  {
+    // If some file operation failed, try to delete the probably-corrupted cache
+    f.Close();
+    File::Delete(filename);
+  }
+  return success;
+}
+
+void GameListCtrl::RescanList()
+{
+  const std::vector<std::string> search_extensions = {".gcm",  ".tgc", ".iso", ".ciso", ".gcz",
+                                                      ".wbfs", ".wad", ".dol", ".elf"};
+  // TODO This could process paths iteratively as they are found
+  auto search_results = Common::DoFileSearch(search_extensions, SConfig::GetInstance().m_ISOFolder,
+                                             SConfig::GetInstance().m_RecursiveISOFolder);
+
+  // TODO rethink some good algorithms to use here
+  std::vector<std::string> cached_paths;
+  for (const auto& file : m_cached_files)
+    cached_paths.emplace_back(file->GetFileName());
+  std::sort(cached_paths.begin(), cached_paths.end());
+
+  std::list<std::string> removed_paths;
+  std::set_difference(cached_paths.cbegin(), cached_paths.cend(), search_results.cbegin(),
+                      search_results.cend(), std::back_inserter(removed_paths));
+
+  std::vector<std::string> new_paths;
+  std::set_difference(search_results.cbegin(), search_results.cend(), cached_paths.cbegin(),
+                      cached_paths.cend(), std::back_inserter(new_paths));
+
+  const Core::TitleDatabase title_database;
+  // TODO we could store all paths and modification times to judge if file needs to be rescanned.
+  // If we cached paths that turned out to be invalid, this would save failing on them each time
+  // refresh is done.
+  // However if people e.g. set dolphin to recursively scan the root of their drive(s), then we
+  // would cache way too much data. Maybe just use an upper bound of invalid paths to cache?
+  // For now, only scan new_paths. This could cause false negatives (file actively being written),
+  // but otherwise
+  // should be fine.
+  for (const auto& path : removed_paths)
+  {
+    auto it = std::find_if(
+        m_cached_files.cbegin(), m_cached_files.cend(),
+        [&path](const std::shared_ptr<GameListItem>& file) { return file->GetFileName() == path; });
+    if (it != m_cached_files.cend())
+      m_cached_files.erase(it);
+  }
+  for (const auto& path : new_paths)
+  {
+    auto file = std::make_shared<GameListItem>(path, title_database);
+    if (file->IsValid())
+      m_cached_files.push_back(std::move(file));
+  }
+
+  for (auto& file : m_cached_files)
+    file->ReloadINI();
```

Comment:
True :( this is a side effect of me wanting to kill off emustate and not being able to do it before this PR. As it is, calling `ReloadINI` on `GameListItem`s that are _already_ shared with the UI thread is technically not safe (although of course in practice it's ok enough for now...). I really hope emustate will be killed and thus the `ReloadINI` call can be removed, which would make synchronizing only the required banners smoother (`ReloadINI` operates on all `m_cached_files`, while `ReloadBannerIfNeeded` only needs to operate on the just-ran wii title or at most all wii titles without banners).